### ISSUE TITLE
BUG: EDICTOR hangs when combining COGIDs that are strings starting with a number

### ISSUE THREAD

**Issue by Cryotops** [digling/edictor#131]
@Anaphory and I just found out that EDICTOR hangs when trying to combine cognate sets if the COGID starts with a number, but is actually a string. 
[edictor.txt](https://github.com/digling/edictor/files/1434127/edictor.txt)

**Comment by LinguList** [comment]
Yes, this is expected behaviour, as you should not use strings as cognate ids. What would you suggest: that we check and convert to numbers automatically upon loading, or rather that I paste with writing a first real tutorial where these things are explained?

**Comment by Anaphory** [comment]
Ah, the confusing thing is that it *seems* to work with non-numerical cognate IDs, *unless* one runs into this. So if you have a good place where you could just add a check for numerical IDs and die (with an error message) if they are not (I haven't found the time to dig into the code to find that place, and @Cryotops is even less at home in JavaScript than I am, so he can't go looking for it), I would consider that sufficient for the moment.

(But then you know that I generally prefer explicit errors to implicit magic, so this shouldn't be a surprise.)

What is the problem with using non-numerical cogids as input? My first thoght was that most obvious way to generate a new cognateset id is `max(COGIDs) + 1`, which gives crazy nonsense for non-numeric ids – but actually, edictor uses the smallest non-assigned positive integer, doesn't it?

**Comment by LinguList** [comment]
Well, you guys are causing me a bit trouble as you are on such a high level of proficiency that you can really tweak the tool to its borders. The average users won't add their own non-numerical ids, as you can't even insert a non-numerical id through the wordlist panel (have you ever tried this?). If you work with the tool and do not use custom-aggregated files that you load into it, the error won't occur. Try it out by typing a question-mark in a wordlist-cogid cell...

I just completed the first draft for a tutorial, online [here](https://github.com/digling/edictor-tutorial) in which I make explicit that a) COGIDs are global, and b) that they are greater zero integers.

> What is the problem with using non-numerical cogids as input? My first thoght was that most obvious way to generate a new cognateset id is max(COGIDs) + 1, which gives crazy nonsense for non-numeric ids – but actually, edictor uses the smallest non-assigned positive integer, doesn't it?

Edictor always searches for the smallest possible unassigned ID. Using non-numerical IDs would leave this to the user (hand-1, etc.), this would be a) local, and b) difficult to increment. Furthermore, when working with the server application, the ID search will even go through the backup and try to find an ID that was not even being used in the past, to make sure that the request for a new ID is ALWAYS a new ID. I know from projects like COBL that non-numerical ids can cause quite some trouble. 

As an alternative, I suggest that you guys look into the possibilities of the MORPHEMES. This allows you to both language-internall and externally assign custom-identifiers, following the logic that everythin that has a space is split. This is good for morophological annotation (provided you have segmented your data morphologically), and it is also useful for some kind of cognate annotation. for the moment, it is only language-internally, but nothing prevents you from using the wordlist-panel to assign cognate-sets manually (you can sort, so you see what is related), and then with lingpy it is as trivial as running ```wordlist.renumber('morphemes', 'cogid'), and a new ID will be created.

I will leave this open and probably add checks later on, but with the two new modules (you may have seen that edictor now offers first support for nexus export via the ANALYZE->Cognate Sets panel), I have to finalize those first. Ideally, I agree, the data should be check when creating the "etymological dictionary", which stores the data underlyingly in JS.

**Comment by Anaphory** [comment]
Well, it's just our data source which is different from what you expect. And if we are told that a certain format is necessary, we can follow up and use that, because then it's our fault that we didn't use the expected style.

Can you tell me what problems COBL had with non-numerical IDs? At the moment, my intuition is that it should be no problem to assume that IDs are arbitrary identifier-strings (No commas in IDs, that would be bad, and probably several other things would also be bad, is that what you mean?).
New cognate IDs generated by Edictor would use the lowest available integer ID – count up from `i=1` and check whether `str(i)` is already used as cognate ID, otherwise assign that.

This intuition may however be wrong, please tell me where it leads to problems.

My mental concept of how edictor might work is also that this is *nearly* what happens at the moment, except that I would look for one `while` or `if` where some implicit conversion between integer and string means that instead of finding `1` as the next free ID, a confusion with `"1a"` leads to an infinite searching loop.
